distributed/shuffle/multi_file.py

```diff
@@ -0,0 +1,293 @@
+    then reads them back when all writes are complete.  It buffers these
+    objects in memory so that it can optimize disk access for larger writes.
+
+    **State**
+
+    -   shards: dict[str, list[bytes]]
+
+        This is our in-memory buffer of data waiting to be sent to other workers.
+
+    -   sizes: dict[str, int]
+
+        The size of each list of shards.  We find the largest and send data from that buffer
+
+    State
+    -----
+
+    memory_limit: str
+        A maximum amount of memory to use, like "1 GiB"
+    max_connections: int
+        The maximum number of connections to have out at once
+    max_message_size: str
+        The maximum size of a single message that we want to send
+
+    Parameters
+    ----------
+    directory: pathlib.Path
+        Where to write and read data.  Ideally points to fast disk.
+    dump: callable
+        Writes an object to a file, like pickle.dump
+    load: callable
+        Reads an object from that file, like pickle.load
+    send: callable
+        How to send a list of shards to a worker
+    sizeof: callable
+        Measures the size of an object in memory
+    """
+
+    memory_limit = parse_bytes("1 GiB")
+    _queues: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+    concurrent_files = 2
+    total_size = 0
+
+    def __init__(
+        self,
+        directory,
+        dump=pickle.dump,
+        load=pickle.load,
+        sizeof=sizeof,
+        loop=None,
+    ):
+        self.directory = pathlib.Path(directory)
+        if not os.path.exists(self.directory):
+            os.mkdir(self.directory)
+        self.dump = dump
+        self.load = load
+        self.sizeof = sizeof
+
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self.total_size = 0
+        self.total_received = 0
+
+        self.condition = asyncio.Condition()
+
+        self.bytes_written = 0
+        self.bytes_read = 0
+
+        self._done = False
+        self._futures = set()
+        self.active = set()
+        self.diagnostics = defaultdict(float)
+
+        self._communicate_future = asyncio.create_task(self.communicate())
+        self._loop = loop or asyncio.get_event_loop()
+        self._exception = None
+
+    @property
+    def queue(self):
+        try:
+            return MultiFile._queues[self._loop]
+        except KeyError:
+            queue = asyncio.Queue()
+            for _ in range(MultiFile.concurrent_files):
+                queue.put_nowait(None)
+            MultiFile._queues[self._loop] = queue
+            return queue
+
+    async def put(self, data: dict[str, list[object]]):
+        """
+        Writes many objects into the local buffers, blocks until ready for more
+
+        Parameters
+        ----------
+        data: dict
+            A dictionary mapping destinations to lists of objects that should
+            be written to that destination
+        """
+        if self._exception:
+            raise self._exception
```

Comment:
Not crucial but I see that flush sets `done` to `False`. IIUC, once we call flush and the buffer is empty, the `communicate` will stop. If after this another `put` happens, we'd just append to the buffer but would never know what happened to it. 

How about raising an exception here if we're done to ensure this object is not accidentally reused? Not reusing this is probably a very important attribute in error cases as well so this would serve two purposes at once.

Alternatively, upon closing we could set the exception attribute with an appropriate exception which would serve a similar purpose, I believe